.github/workflows/build.yml

```diff
@@ -5,21 +5,74 @@ on:
     branches:
       - master
 jobs:
-  check-code-formatting-c-cpp:
-    name: Check code formatting (C/C++)
+  analyze-changes:
+    name: Analyze changes
+    if: github.ref != 'refs/heads/master'
+    runs-on: ubuntu-latest
+    outputs:
+      run-build: ${{steps.change-analysis.outputs.run-build}}
+      run-format-check: ${{steps.change-analysis.outputs.run-format-check}}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2.3.4
+        with:
+          fetch-depth: 0
+      - name: Analyze changes
+        id: change-analysis
+        shell: bash
+        run: >
+          changes=$(
+            git diff --name-only origin/master HEAD -- \
+            ':!*.md' ':!*.asciidoc' ':!.github/CODEOWNERS' ':!externals/*/*' \
+            ':!.github/workflows/rebase.yml' ':!.github/labels.yml' \
+            ':!.github/dependabot.yml' ':!.github/workflows/sync-labels.yml' |
+            wc -l
+          );
+          (( $changes != 0 )) && changes='true' || changes='false';
+          echo "::set-output name=run-build::$changes";
+          echo "Run build: $changes";
+          changes=$(
+            git diff --name-only origin/master HEAD -- \
+            ':!externals/*/*' *.c *.cpp *.h *.hpp *.inl |
+            wc -l
+          );
+          (( $changes != 0 )) && changes='true' || changes='false';
+          echo "::set-output name=run-format-check::$changes";
+          echo "Run format check: $changes"
+  check-code-formatting:
+    name: Check code formatting
+    needs: analyze-changes
+    if:
+      needs.analyze-changes.outputs.run-format-check == 'true' ||
+      github.ref == 'refs/heads/master'
     runs-on: ubuntu-latest
     container: ghcr.io/wnproject/clang-format
     steps:
       - name: Checkout code
         uses: actions/checkout@v2.3.4
+        with:
+          fetch-depth: 0
       - name: Check code format
-        run: >
-          find . -type f
-          -iregex '^.*\.\(c\|cpp\|h\|hpp\|inl\)$' -and -not
-          -iregex '^\.\/externals.*$' |
-          xargs clang-format --dry-run --Werror --verbose
+        run: |
+          if [[ ${{github.ref}} == 'refs/heads/master' ]]; then
```

Comment:
Eventually I want to try and calculate this list as part of the analysis but this works for now.